### PR TITLE
WIP: fsopendir client and server implemetation

### PR DESCRIPTION
**Pull request by masroorhasan**
discussion on fsOpenDir and fsReadDir (on server) helper methods for serialization/deserialization and parsing folder path for client and server.

### CODE REVIEW

File: client/ece454_fs_client.c [hunk 1]
```diff
@@ -96,15 +96,34 @@ extern int fsUnmount(const char *localFolderName) {
...
+               strlen(folderName) + 1,
+               folderName);
+
+    printf("Got response from fsOpenDir RPC.\n");
+    
+    int size = ans.return_size;
+    FSDIR *dir = (FSDIR *) malloc(size);
+    memcpy(dir, (FSDIR *)ans.return_val, size);
+    
+    if(dir != NULL) {
```
Comment (naturalwarren) [inline client/ece454_fs_client.c]: Can we take these prints out before master?

File: server/ece454_fs_server.c [hunk 2]
```diff
@@ -66,6 +89,7 @@ extern return_type fsOpenDir(const int nparams, arg_type *a) {
  */
 extern return_type fsCloseDir(const int nparams, arg_type *a) {
     return_type r;
+    // (void) closedir(dir);
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: Remove.

File: server/ece454_fs_server.c [hunk 3]
```diff
@@ -54,8 +61,24 @@ extern return_type fsUnmount(const int nparams, arg_type *a) {
  * Returns NULL on failures and sets errno.
  */
 extern return_type fsOpenDir(const int nparams, arg_type *a) {
-    return_type r;
-    return r;
+    printf("fsOpenDir() called.\n");
+
+    /*
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: Remove TODO

File: server/ece454_fs_server.c [hunk 4]
```diff
@@ -78,7 +102,34 @@ extern return_type fsCloseDir(const int nparams, arg_type *a) {
...
+    printf("fsReadDir() called.\n");
+
+    int size = sizeof(FSDIR);
+    FSDIR *dir = (FSDIR *) malloc(size);
+    memcpy(dir, (FSDIR *)a->arg_val, size);
+
+    struct dirent *ep;
+
+    if(dir != NULL) {
+        printf("FSDIR in server\n");
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: "Can we change this to printing directory contents."

File: server/ece454_fs_server.c [hunk 5]
```diff
@@ -78,7 +102,34 @@ extern return_type fsCloseDir(const int nparams, arg_type *a) {
...
+        
+        if(readdir((DIR *)dir) != NULL ) {
+
+            while( (ep = readdir(dir))) {
+                printf("%s ", ep->d_name);
+                printf("%d\n", ep->d_namlen);
+            }
+            (void) closedir(dir);
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: I don't think we need a void cast here.

Comment (naturalwarren) [inline server/ece454_fs_server.c]: As discussed just remove this line.

File: server/ece454_fs_server.c [hunk 6]
```diff
@@ -78,7 +102,34 @@ extern return_type fsCloseDir(const int nparams, arg_type *a) {
...
+
+        } else {
+            printf("Couldn't open the directory.\n");
+        }
+        
+    } else {
+        printf("Error in File Directory Stream.\n");
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: Make this "Error reading directory stream."

File: client/ece454_fs_client.c [hunk 7]
```diff
@@ -130,7 +149,18 @@ extern int fsCloseDir(FSDIR * folder) {
...
+    return_type ans;
+    ans = make_remote_call(destAddr,
+               destPort,
+               "fsReadDir", 1,
+               sizeof(FSDIR),
+               (void *)folder);
+
+    printf("Got response from fsReadDir RPC.\n");
+
+
```
Comment (naturalwarren) [inline client/ece454_fs_client.c]: Extra new line here.

File: client/ece454_fs_client.c (hunk 1 above)
Comment (masroorhasan) [inline client/ece454_fs_client.c]: Done.

File: server/ece454_fs_server.c (hunk 2 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Done.

File: server/ece454_fs_server.c [hunk 8]
```diff
@@ -1,5 +1,8 @@
 #include "ece454_fs_server.h"
 
+#include <stdlib.h>
```
Comment (naturalwarren) [inline server/ece454_fs_server.c]: Includes should always go in the header file.

File: server/ece454_fs_server.c (hunk 3 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Done.

File: server/ece454_fs_server.c (hunk 4 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Changed to "Printing directory contents."

File: server/ece454_fs_server.c (hunk 5 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Done.

File: server/ece454_fs_server.c (hunk 6 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Done.

File: client/ece454_fs_client.c (hunk 7 above)
Comment (masroorhasan) [inline client/ece454_fs_client.c]: Done.

File: server/ece454_fs_server.c (hunk 8 above)
Comment (masroorhasan) [inline server/ece454_fs_server.c]: Done.

File: server/ece454_fs_server.h [hunk 9]
```diff
@@ -13,6 +13,8 @@
 #include <sys/types.h>
 #include <dirent.h>
 
+#include <string.h>
```
Comment (naturalwarren) [inline server/ece454_fs_server.h]: Put with system imports.

**Pull request merged**